community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/ImportLogic.java

```diff
@@ -0,0 +1,528 @@
+     */
+    public ImportLogic( File storeDir, FileSystemAbstraction fileSystem, BatchingNeoStores neoStore,
+            Configuration config, LogService logService, ExecutionMonitor executionMonitor,
+            RecordFormats recordFormats, Input input )
+    {
+        this.storeDir = storeDir;
+        this.fileSystem = fileSystem;
+        this.neoStore = neoStore;
+        this.config = config;
+        this.recordFormats = recordFormats;
+        this.log = logService.getInternalLogProvider().getLog( getClass() );
+        this.executionMonitor = ExecutionSupervisors.withDynamicProcessorAssignment( executionMonitor, config );
+        this.maxMemory = config.maxMemoryUsage();
+
+        initialize( input );
+    }
+
+    private void initialize( Input input )
+    {
+        log.info( "Import starting" );
+        startTime = currentTimeMillis();
+        inputCache = new InputCache( fileSystem, storeDir, recordFormats, config );
+
+        numberArrayFactory = auto( neoStore.getPageCache(), storeDir, config.allowCacheAllocationOnHeap() );
+        badCollector = input.badCollector();
+        // Some temporary caches and indexes in the import
+        writeMonitor = new IoMonitor( neoStore.getIoTracer() );
+        idMapper = input.idMapper( numberArrayFactory );
+        idGenerator = input.idGenerator();
+        nodeRelationshipCache = new NodeRelationshipCache( numberArrayFactory, config.denseNodeThreshold() );
+        memoryUsageStats = new MemoryUsageStatsProvider( nodeRelationshipCache, idMapper );
+        nodes = input.nodes();
+        relationships = input.relationships();
+        cachedNodes = cachedForSure( nodes, inputCache.nodes( MAIN, true ) );
+    }
+
+    /**
+     * Accesses state of a certain {@code type}. This is state that may be long- or short-lived and perhaps
+     * created in one part of the import to be used in another.
+     *
+     * @param type {@link Class} of the state to get.
+     * @return the state of the given type.
+     * @throws IllegalStateException if the state of the given {@code type} isn't available.
+     */
+    public <T> T getState( Class<T> type )
+    {
+        return type.cast( accessibleState.get( type ) );
+    }
+
+    /**
+     * Puts state of a certain type.
+     *
+     * @param state state instance to set.
+     * @see #getState(Class)
+     * @throws IllegalStateException if state of this type has already been defined.
+     */
+    public <T> void putState( T state )
+    {
+        accessibleState.put( state.getClass(), state );
+    }
+
+    /**
+     * Imports nodes w/ their properties and labels from {@link Input#nodes()}. This will as a side-effect populate the {@link IdMapper},
+     * to later be used for looking up ID --> nodeId in {@link #importRelationships()}. After a completed node import,
+     * {@link #prepareIdMapper()} must be called.
+     *
+     * @throws IOException on I/O error.
+     */
+    public void importNodes() throws IOException
```

Comment:
importNodes and importRelationships steps perfrom: 
neoStore.startFlushingPageCache(); 
execute... 
neoStore.stopFlushingPageCache();
why do we need those only for those 2 steps and why we can't have background flusher that flushes data during any step executuion?

Reply:
It turns out that the background flushing helps some stages and makes some worse. There are also some stages that don't write anything at all.